src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java

```diff
@@ -0,0 +1,217 @@
+        return objectNode;
+    }
+
+    private Publisher makePublisher(BidRequest bidRequest, int publisherId) {
+        final Publisher publisher = bidRequest.getSite() != null
+                ? bidRequest.getSite().getPublisher()
+                : bidRequest.getApp().getPublisher();
+
+        final Publisher.PublisherBuilder publisherBuilder = publisher == null
+                ? Publisher.builder()
+                : publisher.toBuilder();
+
+        return publisherBuilder
+                .id(String.valueOf(publisherId))
+                .build();
+    }
+
+    private Integer getPublisherId(Map<String, ExtImpUndertone> extImpMap) {
+        return extImpMap.values()
+                .stream()
+                .map(ExtImpUndertone::getPublisherId)
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElse(null);
+    }
+
+    private List<Imp> makeImps(BidRequest bidRequest, Map<String, ExtImpUndertone> extImpMap) {
+        return bidRequest.getImp()
+                .stream()
+                .filter(imp -> isValidImp(imp, extImpMap.get(imp.getId())))
+                .map(imp -> imp.toBuilder()
+                        .tagid(extImpMap.get(imp.getId()).getPlacementId().toString())
+                        .ext(null)
+                        .build())
+                .toList();
+    }
+
+    private Map<String, ExtImpUndertone> getUndertoneExtImpMap(BidRequest bidRequest) {
+        return bidRequest.getImp()
+                .stream()
+                .collect(Collectors.toUnmodifiableMap(Imp::getId,
+                        imp -> mapper.mapper().convertValue(imp.getExt(), UNDERTONE_EXT_TYPE_REFERENCE).getBidder()));
+    }
+
+    private boolean isValidImp(Imp imp, ExtImpUndertone extImpUndertone) {
+        return imp != null && ObjectUtils.anyNotNull(imp.getVideo(), imp.getBanner())
+                && extImpUndertone != null && extImpUndertone.getPlacementId() != null;
+    }
+
+    private List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return List.of();
+        }
+        return bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
+        final Map<String, Imp> idImpMap = getIdImpMap(bidRequest);
+
+        return bidResponse.getSeatbid()
+                .stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid, getBidType(bid, idImpMap), bidResponse.getCur()))
+                .toList();
+    }
+
+    private Map<String, Imp> getIdImpMap(BidRequest bidRequest) {
+        return bidRequest.getImp()
+                .stream()
+                .collect(Collectors.groupingBy(Imp::getId))
```

Comment:
Core will not let you have two imps with identical id.
This code is redundant

Reply:
`getBidType` method uses this map, would you like us to change the implementation?